hummingbird/ml/_container.py

```diff
@@ -478,16 +465,86 @@ def _decision_function(self, *inputs):
+        assert tvm_installed()
+        import tvm
+
+        self._ctx = self._extra_config[constants.TVM_CONTEXT]
+        self._input_names = self._extra_config[constants.TVM_INPUT_NAMES]
+        self._remainder_model = None
+        if constants.TVM_REMAINDER_MODEL in self._extra_config:
+            self._remainder_model = self._extra_config[constants.TVM_REMAINDER_MODEL]
+        self._to_tvm_array = lambda x: tvm.nd.array(x, self._ctx)
+
+        os.environ["TVM_NUM_THREADS"] = str(self._n_threads)
+
+    def _to_tvm_tensor(self, *inputs):
+        return {self._input_names[i]: self._to_tvm_array(inputs[i]) for i in range(len(inputs))}
+
+
+class TVMSklearnContainerTransformer(TVMSklearnContainer, SklearnContainerTransformer):
+    """
+    Container for TVM models mirroring Sklearn transformers API.
+    """
+
+    def _transform(self, *inputs):
+        if self._last_iteration and self._remainder_model is not None:
+            self._remainder_model.run(**self._to_tvm_tensor(*inputs))
+            return self._remainder_model.get_output(0).asnumpy()
+        self.model.run(**self._to_tvm_tensor(*inputs))
+        return self.model.get_output(0).asnumpy()
+
+
+class TVMSklearnContainerRegression(TVMSklearnContainer, SklearnContainerRegression):
+    """
+    Container for TVM models mirroring Sklearn regressors API.
+    """
+
+    def _predict(self, *inputs):
+        if self._last_iteration and self._remainder_model is not None:
+            self._remainder_model.run(**self._to_tvm_tensor(*inputs))
+            return self._remainder_model.get_output(0).asnumpy().ravel()
+        self.model.run(**self._to_tvm_tensor(*inputs))
+        return self.model.get_output(0).asnumpy().ravel()
+
+
+class TVMSklearnContainerClassification(TVMSklearnContainerRegression, SklearnContainerClassification):
+    """
+    Container for TVM models mirroring Sklearn classifiers API.
+    """
+
+    def _predict_proba(self, *inputs):
+        if self._last_iteration and self._remainder_model is not None:
+            self._remainder_model.run(**self._to_tvm_tensor(*inputs))
+            return self._remainder_model.get_output(1).asnumpy()
+        self.model.run(**self._to_tvm_tensor(*inputs))
+        return self.model.get_output(1).asnumpy()
+
+
+class TVMSklearnContainerAnomalyDetection(TVMSklearnContainerRegression, SklearnContainerAnomalyDetection):
+    """
+    Container for TVM models mirroring Sklearn anomaly detection API.
+    """
+
+    def _decision_function(self, *inputs):
+        if self._last_iteration and self._remainder_model is not None:
+            self._remainder_model.run(**self._to_tvm_tensor(*inputs))
+            scores = self._remainder_model.get_output(1).asnumpy().ravel()
+        else:
+            self.model.run(**self._to_tvm_tensor(*inputs))
+            scores = self.model.get_output(1).asnumpy().ravel()
+
+        # Backward compatibility for sklearn <= 0.21
+        if constants.IFOREST_THRESHOLD in self._extra_config:
```

Comment:
These 2 lines now appear in 3 places. consider to combine/make function?

Reply:
I removed one occurrence, that was actually a bug.